internal/remotewrite/proxy.go

```diff
@@ -0,0 +1,174 @@
+	Endpoints []Endpoint `yaml:"endpoints"`
+}
+
+type Endpoint struct {
+	Name string `yaml:"name"`
+	URL  string `yaml:"url"`
+	// +optional
+	TlsConfig *TlsConfig `yaml:"tlsConfig,omitempty"`
+	// +optional
+	BasicAuth *BasicAuth `yaml:"basicAuth,omitempty"`
+}
+
+type TlsConfig struct {
+	CA   string `yaml:"ca"`
+	Cert string `yaml:"cert"`
+	Key  string `yaml:"key"`
+}
+
+type BasicAuth struct {
+	User     string `yaml:"user"`
+	Password string `yaml:"password"`
+}
+
+func Proxy(write *url.URL, endpoints *Endpoints, logger log.Logger, r *prometheus.Registry) http.Handler {
+
+	requests := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name:        "http_proxy_requests_total",
+		Help:        "Counter of proxy HTTP requests.",
+		ConstLabels: prometheus.Labels{"proxy": "metricsv1-write"},
+	}, []string{"method"})
+	r.MustRegister(requests)
+
+	if endpoints == nil {
+		endpoints = &Endpoints{}
+	}
+
+	if write != nil {
+		endpoints.Endpoints = append(endpoints.Endpoints, Endpoint{
+			URL:  write.String(),
+			Name: THANOS_ENDPOINT_NAME,
+		})
+	}
+
+	remotewriteRequests := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name:        "remote_write_requests_total",
+		Help:        "Counter of remote write requests.",
+		ConstLabels: prometheus.Labels{"proxy": "metricsv1-remotewrite"},
+	}, []string{"code", "name"})
+	r.MustRegister(remotewriteRequests)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		requests.With(prometheus.Labels{"method": r.Method}).Inc()
+
+		body, _ := ioutil.ReadAll(r.Body)
+		_ = r.Body.Close()
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+		if endpoints == nil {
+			endpoints = &Endpoints{}
+		}
+		if write != nil {
+			remotewriteUrl := url.URL{}
+			remotewriteUrl.Path = path.Join(write.Path, r.URL.Path)
+			remotewriteUrl.Host = write.Host
+			remotewriteUrl.Scheme = write.Scheme
+			endpoints.Endpoints[len(endpoints.Endpoints)-1].URL = remotewriteUrl.String()
+		}
+
+		rlogger := log.With(logger, "request", middleware.GetReqID(r.Context()))
+		for _, endpoint := range endpoints.Endpoints {
+			client := &http.Client{}
+			if endpoint.TlsConfig != nil {
+				transport, err := createMtlsTransport(*endpoint.TlsConfig)
+				if err != nil {
+					level.Error(rlogger).Log("msg", "Failed to create mtls transport", "err", err, "url", endpoint.URL)
+					return
+				}
+				client.Transport = transport
+			}
+			req, err := http.NewRequest(http.MethodPost, endpoint.URL, bytes.NewReader(body))
+			req.Header = r.Header
+			if err != nil {
+				level.Error(rlogger).Log("msg", "Failed to create the forward request", "err", err, "url", endpoint.URL)
+			} else {
+				if endpoint.BasicAuth != nil {
+					auth := fmt.Sprintf("%s:%s", endpoint.BasicAuth.User, endpoint.BasicAuth.Password)
+					authHeader := fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(auth)))
+					req.Header.Add("Authorization", authHeader)
+				}
+				ep := endpoint
+				go func() {
```

Comment:
will the goroutine be created for each request? in my understand, each request triggers the HandlerFunc function, then create this goroutine. correct?

Reply:
right. do you think this will lead to some perf problems?

Reply:
it is possible. Why we cannot have one dedicated go routine for one endpoint?

Reply:
then, there maybe tons of requests stuck in the dedicated go routine because the requests to the same endpoint will be sent out one by one by the go routine.
I considered to use something like thread pool before. But seems go routine does not require that since it's very lightweight